File: 0024-swap-nodes-in-pairs/0024-swap-nodes-in-pairs.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var swapPairs = function (head) {
    if (!head) return head;

    const dummy = new ListNode();
    let prev = dummy;
    prev.next = head;

    let curr = head;

    while (curr) {
        let next = curr.next;

        if (next) {
            curr.next = next.next;
            next.next = curr;
            prev.next = next;
        }

        prev = curr;
        curr = curr.next;
    }

    return dummy.next;
};

File: 0024-swap-nodes-in-pairs/0024-swap-nodes-in-pairs.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapPairs(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return head

        dummy = ListNode()
        prev = dummy

        prev.next = head
        curr = head

        while curr:
            next = curr.next

            if next:
                curr.next = next.next
                next.next = curr
                prev.next = next
            
            prev = curr
            curr = curr.next
        return dummy.next